### ISSUE TITLE
calculate all q\k linear attention？

### ISSUE THREAD

**Issue by TribleH** [thu-ml/SLA#15]
This code use all q,k calculate linear attention,But in the paper, it needs to be calculated according to the M matrix, and it is only calculated when the M matrix is 0。

<img width="1132" height="610" alt="Image" src="https://github.com/user-attachments/assets/b2ed87fd-4b7e-4ce3-8069-39a4916c08f3" />

def calc_linear(q, k, v):
            kvsum = k.transpose(-1, -2) @ v
            ksum = torch.sum(k, dim=-2, keepdim=True)
            return (q @ kvsum) / (1e-5 + (q * ksum).sum(dim=-1, keepdim=True))
        o_l = calc_linear(c_q, c_k, v)

        with torch.amp.autocast('cuda', dtype=self.dtype):
            o_proj = self.proj_l(o_l)
        o = (o_s + o_proj).to(dtype)

**Comment by jt-zhang** [comment]
We update the code in the main branch. The original code is in the SLA-1 branch.

You can use one of them.
